src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,936 @@
+"""
+Define a BlkDiagMatrix module which implements operations for
+block diagonal matrices as used by ASPIRE.
+"""
+
+import numpy as np
+
+from numpy.linalg import norm
+from numpy.linalg import solve
+from scipy.linalg import block_diag
+from scipy.special import jv
+
+from aspire.utils.cell import Cell2D
+from aspire.basis.ffb_2d import FFBBasis2D
+from aspire.basis.basis_utils import lgwt
+
+
+class BlkDiagMatrix:
+    """
+    Define a BlkDiagMatrix class which implements operations for
+    block diagonal matrices as used by ASPIRE.
+
+    Currently BlkDiagMatrix is implemented only for square blocks.
+    While in the future this can be extended, at this time assigning
+    a non square array will raise NotImplementedError.
+    """
+
+    # Developers' Note:
+    # All instances of this class should have priority over ndarray ops
+    #   because we implement them here ourselves.
+    # This is a more np current implementation of __array_priority__
+    #   operator precedence schedule.
+    # Mainly this effects rmul, radd, rsub eg:
+    #   blk_y = scalar_a + ( scalar_b * blk_x)
+    __array_ufunc__ = None
+
+    def __init__(self, partition, dtype=np.float64):
+        """
+        Instantiate a BlkDiagMatrix.
+
+        :param partition: The matrix block partition
+         in the form of a `nblock`-element list storing all shapes of
+         diagonal matrix blocks, where `partition[i]` corresponds to
+         the shape (number of rows and columns) of the `i` matrix block.
+        :param dtype: Datatype for blocks, defaults to np.float64.
+        :return: BlkDiagMatrix instance.
+        """
+
+        self.nblocks = len(partition)
+        self.dtype = np.dtype(dtype)
+        self.data = [None] * self.nblocks
+        self._cached_blk_sizes = np.array(partition)
+        if len(partition):
+            assert self._cached_blk_sizes.shape[1] == 2
+            assert all([BlkDiagMatrix.check_square(shp) for shp in partition])
+
+    @staticmethod
+    def check_square(shp):
```

Comment:
Make private function. Does this need to be in the class at all? Seems more like a general utility function.